spoc/snipping/snipping_strategies.py

```diff
@@ -0,0 +1,240 @@
+"""Snipping strategies for Snipper that implement specific snipping functionality"""
+from abc import ABC, abstractmethod
+from functools import partial
+from enum import Enum
+from typing import Union
+from multiprocess.pool import ThreadPool
+import pandas as pd
+import numpy as np
+from sparse import COO
+import duckdb
+import bioframe
+from spoc.pixels import PersistedPixels
+
+
+class SnippingValues(Enum):
+    """Which values the snippet should consist of."""
+
+    ICCF = 0
+    OBSEXP = 1
+
+
+class SnippingStrategy(ABC):
+    """Defines interface for snipping strategy"""
+
+    def __init__(
+        self,
+        bin_size: int,
+        half_window_size: int,
+        snipping_value: Union[str, SnippingValues],
+        **kwargs,
+    ):
+        """Defines the values that should be snipped"""
+        if isinstance(snipping_value, str):
+            # check whether string refers to a snipping strategy
+            snipping_value = SnippingValues[snipping_value.upper()]
+        self._snipping_value = snipping_value
+        self._half_window_size = half_window_size
+        if snipping_value == SnippingValues.OBSEXP:
+            self._n_random_regions: int = kwargs.get("n_random_regions", 5000)
+            self._genome: str = kwargs.get("genome", "hg19")
+        self._bin_size = bin_size
+
+    @staticmethod
+    def _get_random_coordinates(n_coordinates: int, length: int, genome: str):
+        """Number of coordinates will not be exactly returned, due to rounding
+        when distributing to chromosomes."""
+        chrom_sizes = bioframe.fetch_chromsizes(genome)
+        chrom_fractions = chrom_sizes / chrom_sizes.sum()
+        # accumulate output
+        chrom_frames = []
+        for chrom in chrom_sizes.index:
+            max_coord = chrom_sizes[chrom]
+            number_to_choose = int(chrom_fractions[chrom] * n_coordinates)
+            starts = np.random.randint(0, max_coord, size=number_to_choose)
+            ends = starts + length
+            chrom_frames.append(
+                pd.DataFrame({"chrom": chrom, "start": starts, "end": ends})
+            )
+        return pd.concat(chrom_frames)
+
+    def __repr__(self) -> str:
+        return f"<Triplet1DSnippingStrategy>"
+
+    @abstractmethod
+    def get_params():
+        raise NotImplementedError
+
+    @abstractmethod
+    def snip(
+        self,
+        pixels: Union[pd.DataFrame, PersistedPixels],
+        snip_positions: pd.DataFrame,
+        threads: int = 2,
+        override_snipping_value: bool = False,
+    ) -> pd.DataFrame:
+        """Do snipping"""
+        raise NotImplementedError
```

Comment:
No need to raise an Error if the method is already decorated with `abstractmethod`. Python will raise an error unless all abstract methods are overridden.